<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Paul Villanueva" />

<meta name="date" content="2019-10-15" />

<title>Plotting coverage</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 54px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h2 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h3 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h4 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h5 {
  padding-top: 59px;
  margin-top: -59px;
}
.section h6 {
  padding-top: 59px;
  margin-top: -59px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Index</a>
</li>
<li>
  <a href="elbows.html">Elbows</a>
</li>
<li>
  <a href="ggcoverage.html">Plotting BLAST Coverage</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Plotting coverage</h1>
<h4 class="author">Paul Villanueva</h4>
<h4 class="date">10/15/2019</h4>

</div>


<div id="plotting-coverage-from-blast-results-using-ggplot2" class="section level1">
<h1>Plotting coverage from BLAST results using ggplot2</h1>
<div id="preparing-the-data" class="section level2">
<h2>Preparing the data</h2>
<p>The data we’re working with is output from using BLAST against a single gene, where the queries are the metagenomes we’re interested in. Assuming you used BLAST with <code>-outfmt 6</code>, extract the <code>sstart</code> and <code>send</code> columns from the <code>blast_results</code> file and save them to a new file by:</p>
<pre class="bash"><code>cut -f 9,10 blast_results &gt; regions_hit.tsv</code></pre>
<p>Your file should look like this:</p>
<pre class="bash"><code>head data/tet33.regions_hit.tsv</code></pre>
<pre><code>## 969  820
## 161  12
## 46   103
## 923  1066
## 722  573
## 293  144
## 327  178
## 173  24
## 861  745
## 1005 856</code></pre>
<p>Then read the data into a dataframe with:</p>
<pre class="r"><code>library(tidyverse)
regions_hit &lt;- read.delim(&quot;data/tet33.regions_hit.tsv&quot;,
                          col.names = c(&quot;sstart&quot;, &quot;send&quot;)) %&gt;% 
  mutate(forward = ifelse(sstart &gt; send, 0, 1))
head(regions_hit)</code></pre>
<pre><code>##   sstart send forward
## 1    161   12       0
## 2     46  103       1
## 3    923 1066       1
## 4    722  573       0
## 5    293  144       0
## 6    327  178       0</code></pre>
<p>The mutate statement above adds an extra column indicating whether or not the match was on the reverse strand.</p>
</div>
<div id="plotting-the-coverage" class="section level2">
<h2>Plotting the coverage</h2>
<p>Now for the function itself:</p>
<pre class="r"><code>plot.coverage &lt;- function(regions_hit.df, roi = NULL){
  library(IRanges)
  library(ggplot2)
  
  x.min = min(regions_hit.df[, 1:2])
  x.max = max(regions_hit.df[, 1:2])
  
  forward_hits &lt;- regions_hit.df[regions_hit.df$forward == 1, ]
  
  forward_hits.IR = IRanges(start = forward_hits[, 1],
                            end = forward_hits[, 2])
  forward_hits.cov &lt;- coverage(forward_hits.IR)
  
  forward_coverage &lt;- data.frame(pos = seq(x.min, x.max),
                                 cov = unlist(mapply(rep, forward_hits.cov@values, forward_hits.cov@lengths)),
                                 forward = as.factor(rep(1, x.max))
                      )
  
  reverse_hits &lt;- regions_hit.df[regions_hit.df$forward == 0, ]
  reverse_hits.IR = IRanges(start = reverse_hits[, 2],
                            end = reverse_hits[, 1])
  
  reverse_hits.cov &lt;- coverage(reverse_hits.IR)
  reverse_coverage &lt;- data.frame(pos = seq(x.min, x.max),
                                 cov = unlist(mapply(rep, reverse_hits.cov@values, reverse_hits.cov@lengths)),
                                 forward = as.factor(rep(0, x.max))
                      )
  
  total_coverage &lt;- rbind(forward_coverage, reverse_coverage)
  
  y.max &lt;-max(total_coverage$cov)

  p &lt;- ggplot(total_coverage, aes(x = pos, y = cov, color = forward)) +
    geom_line() +
    geom_hline(yintercept = 0) + 
    scale_x_continuous(expand = c(0, x.min), limits = c(0, x.max)) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, y.max * 1.1)) +
    labs(x = &quot;Position&quot;, 
         y = &quot;Depth&quot;,
         title = &quot;Coverage depth of...&quot;,
         color = &quot;Strand&quot;) + 
    scale_color_manual(labels = c(&quot;Forward&quot;, &quot;Reverse&quot;), 
                       values = c(&quot;#FF6633&quot;, &quot;#99CC99&quot;)) +
    theme_minimal() +
    theme(panel.border = element_rect(color = &quot;black&quot;, fill = NA, size = 1))
  
  if (!missing(roi)){
    p &lt;- p + geom_vline(xintercept = roi, linetype = &quot;dashed&quot;)
  }
  
  return(p)
}</code></pre>
<p>The <code>plot.coverage</code> function takes as input a dataframe with columns indicating the start, end, and whether or not we’re on the forward or reverse strand. For example:</p>
<pre class="r"><code>plot.coverage(regions_hit)</code></pre>
<p><img src="ggcoverage_files/figure-html/example.1-1.png" width="672" /></p>
<p><code>plot.coverage</code> returns a <code>ggplot</code> object, so you can make any modifications you’d like to the graph in the usual way:</p>
<pre class="r"><code>plot.coverage(regions_hit) + 
  labs(title = &quot;Coverage depth of tet33 in soil column 1 metagenomes&quot;)</code></pre>
<p><img src="ggcoverage_files/figure-html/example.2-1.png" width="672" /></p>
<p>Additionally, <code>plot.coverage</code> takes an optional <code>roi</code> parameter to indiciate a <em>r</em>egion <em>o</em>f <em>i</em>nterest:</p>
<pre class="r"><code>plot.coverage(regions_hit, c(300, 567)) + 
  labs(title = &quot;Coverage depth of tet33 in soil column 1 metagenomes&quot;,
       subtitle = &quot;Number of reads per position; dashed lines indicate region targeted by primer pair&quot;)</code></pre>
<p><img src="ggcoverage_files/figure-html/example.3-1.png" width="672" /></p>
<p>In the future, I will rewrite the function to use <code>data.table</code> and add an optional argument to separate the forward and reverse reads across the <span class="math inline">\(x\)</span> axis.</p>
</div>
</div>

<p><br>
<br>
<strong><a href="https://github.com/pommevilla">Paul Villanueva</a></strong> 
<br>
Ph.D. Student - Bioinformatics and Computational Biology<br>
Iowa State University, Ames, IA.</p>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
